File: python/algos/majority_element.py
def majority_element(numbers: list[int]) -> int:
    score = 0
    candidate = None
    for value in numbers:
        if score == 0:
            candidate = value
            score = 1
        elif value == candidate:
            score += 1
        else:
            score -= 1
    return candidate


def test_majority_element():
    assert majority_element([3, 2, 3]) == 3
    assert majority_element([2, 2, 1, 1, 1, 2, 2]) == 2

File: go/majority_element.go
package algos

func majorityElement(numbers []int) int {
	score := 0
	var candidate int

	for i := 0; i < len(numbers); i++ {
		if score == 0 {
			score = 1
			candidate = numbers[i]
		} else if numbers[i] == candidate {
			score += 1
		} else {
			score -= 1
		}
	}
	return candidate
}
